documentation/documentation_publisher.ts

```diff
@@ -0,0 +1,132 @@
+/* eslint-disable no-console */
+import AWS from 'aws-sdk';
+import {Command} from 'commander';
+import S3 from 'aws-sdk/clients/s3';
+import * as fs from 'fs';
+import * as mime from 'mime';
+import * as path from 'path';
+
+const AwsRegion = 'us-west-2';
+const BaseGeneratedDocsPath = 'site';
+const DocumentationBucket = 'developer-documentation';
+const PacksSdkBucketRootPath = 'packs';
+const EnvironmentKeyRoot = `${PacksSdkBucketRootPath}/current`;
+const Environments = ['adhoc', 'head', 'staging', 'prod'];
+const program = new Command();
+
+// only works after `program` has been parsed
+function isVerbose() {
+  return program.opts().verbose;
+}
+
+function handleError(e: Error) {
+  console.error(e);
+  process.exit(1);
+}
+
+function getS3Service(env: string): S3 {
+  const config: S3.ClientConfiguration = {
+    signatureVersion: 'v4',
+    region: AwsRegion,
+    credentials: new AWS.SharedIniFileCredentials({profile: env}),
+    computeChecksums: true,
+  };
+  return new S3(config);
+}
+
+function getS3Bucket(env: string): string {
+  return `coda-us-west-2-${env}-${DocumentationBucket}`;
+}
+
+function getS3ConfigKey(hash: string): string {
+  return `${PacksSdkBucketRootPath}/${hash}`;
+}
+
+function getS3EnvironmentKey(environment: string): string {
+  return `${EnvironmentKeyRoot}/${environment}`;
+}
+
+async function pushDocsToEnv(env: string, version: string) {
+  const s3 = getS3Service(env);
+  const bucket = getS3Bucket(env);
+  const key = getS3ConfigKey(version);
+
+  async function pushDocsDirectory(basePath: string): Promise<any> {
+    const fullFilePath = path.join(BaseGeneratedDocsPath, basePath);
+    const fileStats = fs.lstatSync(fullFilePath);
+    if (fileStats.isFile()) {
+      const contentType = mime.getType(fullFilePath) || undefined;
+      const fileData = fs.createReadStream(fullFilePath);
+      if (isVerbose()) {
+        console.log(`Uploading ${basePath} to S3 as a file`);
+      }
+      return s3
+        .putObject({
+          Bucket: bucket,
+          Key: `${key}/${basePath}`,
+          Body: fileData,
+          ContentType: contentType,
+        })
+        .promise();
+    }
+
+    if (fileStats.isDirectory()) {
+      if (isVerbose()) {
+        console.log(`Recursively traversing the directory ${basePath}.`);
+      }
+      const files = fs.readdirSync(fullFilePath);
+      return Promise.all(
+        files.map(fileName => {
+          const nestedFilePath = path.join(basePath, fileName);
+          return pushDocsDirectory(nestedFilePath);
+        }),
+      );
+    }
+  }
+
+  try {
+    console.log(`Pushing the current packs-sdk documentation ${key} to ${env}...`);
```

Comment:
I believe we also have a `print` helper, but if your code isn't triggering our no-console lint rule then it's probably fine to leave as is?